### PR TITLE
Add SwiftUI enhancements

### PR DESCRIPTION
**Pull request by jsflax**
This PR adds the `@StateRealmObject`, `@ObservedObject`, and `@ObservedResults` property wrappers, which will allow users to more easily wield thread confined live types: `Objects`, `EmbeddedObjects`, `Lists`, or `Results`:

```swift
struct PersonView: View {
    @ObservedResults(Person.self) var results

    var body: some View {
        return NavigationView {
            List {
                ForEach(results) { person in
                    NavigationLink(destination: PersonDetailView(person: person)) {
                        Text(person.name)
                    }
                }
                .onDelete(perform: $results.remove)
            }
            .navigationBarTitle("People", displayMode: .large)
            .navigationBarItems(trailing: Button("Add") {
                $results.append(Person())
            })
        }
    }
}
```

The wrapper allows properties to be wielded normally when doing most operations, and adds bonus convenience methods for under-the-hood write transactions for each type (and their properties).

### CODE REVIEW

File: RealmSwift/SwiftUI.swift
Comment (jsflax) [reply RealmSwift/SwiftUI.swift:67]: The setter cannot throw because it is used to set the wrappedValue with the nonmutating setter. What's the alternative?

File: SwiftUITests/SwiftUITests.swift
Comment (leemaguire) [inline SwiftUITests/SwiftUITests.swift:1]: needs a licence header

File: examples/ios/swift/ListSwiftUI/Views/ContentView.swift
Comment (leemaguire) [inline examples/ios/swift/ListSwiftUI/Views/ContentView.swift:68]: will this `bind` API be hard to discover? Do we also want to extend `View` to do the also?

Comment (jsflax) [reply examples/ios/swift/ListSwiftUI/Views/ContentView.swift:68]: My other thought was to put `bind` on `ThreadConfined` types, but I'm not sure if there is value in that outside of SwiftUI (it's API bloat otherwise, but maybe that's fine?). I'm open to ideas anyways.

Comment (leemaguire) [reply examples/ios/swift/ListSwiftUI/Views/ContentView.swift:68]: I'm just thinking we should extend our own domain and not another party's. But it's tricky because its SwiftUI specific. I guess I'm fine with extending `View` as its a very niche case

Comment (jsflax) [reply examples/ios/swift/ListSwiftUI/Views/ContentView.swift:68]: I actually don't think this is terribly niche, which is why I'm trying to be careful– binding to a SwiftUI view will likely be fairly common (TextFields, sliders, etc.).

Thoughts @tgoyne ?

**Comment by jsflax** [comment]
@tgoyne I've diverged from Eric's branch and will changeover to master once Eric's changes are merged, but could use another set of eyes on the actual swiftUI related changes.

File: Configuration/SwiftUITestHost.xcconfig
Comment (tgoyne) [inline Configuration/SwiftUITestHost.xcconfig:13]: This needs to target 10.15 to be able to run on CI.

File: Realm.xcodeproj/project.pbxproj
Comment (tgoyne) [inline Realm.xcodeproj/project.pbxproj:2864]: This is why you have to do awkward things with code signing and it should just be off unless it's actually required.

File: RealmSwift/List.swift
Comment (tgoyne) [inline RealmSwift/List.swift:621]: Is there some reason not to use this code path for non-embedded objects as well?

File: RealmSwift/SwiftUI.swift
Comment (tgoyne) [inline RealmSwift/SwiftUI.swift:34]: Can't this just be `parent.realm!.thaw(parent)`?

The force-unwrap is unsafe and this'll crash if the object was deleted after the version which the parent was frozen at.

File: RealmSwift/SwiftUI.swift
Comment (tgoyne) [inline RealmSwift/SwiftUI.swift:156]: get() can do some nontrivial work so this should avoid calling it twice in one function.

File: RealmSwift/SwiftUI.swift
Comment (tgoyne) [inline RealmSwift/SwiftUI.swift:334]:
```suggestion
        let actualRealm = realm ?? try! Realm()
```

File: RealmSwift/List.swift
Comment (jsflax) [reply RealmSwift/List.swift:621]: It seemed like the way for non-embedded objects was "more correct," and that this was a workaround, but I can check if that actually matters

File: RealmSwift/SwiftUI.swift
Comment (iandundas) [inline RealmSwift/SwiftUI.swift:594]: small typo:
```suggestion
                // if unmanaged, return the unmanaged value
```

**Comment by jsflax** [comment]
Some changes since last review:
- KVO has been added for unmanaged objects
- Optionals are now supported for `@StateRealmObject`
- `@StateRealmObject` is now exclusively for `List`, `Objects`/`EmbeddedObjects`, and `Optional` variants of the previously mentioned types. It is meant to be used in the same way as SwiftUI's `@StateObject`.
- `@ObservedRealmObject` is now exclusively for `List`, `Objects`/`EmbeddedObjects`. It is meant to be used in the same way as SwiftUI's `@ObservedObject`.
- `@FetchRealmResults` has been added to better support binding Results types to a view while allowing dynamic querying and multiple Realm support. The default realm configuration for a `@FetchRealmResults` is passed in through the environment of a view hierarchy, e.g.,:
```swift
struct View2: View {
    // the realm used here will be the realm opened from
    // `myCustomConfiguration`
    @FetchRealmResults(MyObject.self) var results

    ...
}

struct View1: View {
    var body: some View {
        View2().environment(\.realmConfiguration, myCustomConfiguration)
    }
}
```
- The example in the examples directory has been changed to the Reminders application

TODO:
- Need to update the README.
- Add one example showing multiple realm usage

@tgoyne can you PTAL?

File: RealmSwift/SwiftUI.swift
Comment (jsflax) [reply RealmSwift/SwiftUI.swift]: Forgot to, but I still need to do this even though this comment is outdated.

File: RealmSwift/List.swift
Comment (jsflax) [reply RealmSwift/List.swift]: I've added a test for this and collapsed the conditional.

**Comment by jsflax** [comment]
@tgoyne I've resolved all previous comments, just wrestling with the CI now.

File: CHANGELOG.md
Comment (leemaguire) [inline CHANGELOG.md:7]: should this be `@ObservedRealmObject`?

Comment (jsflax) [reply CHANGELOG.md:7]: Yes

File: CHANGELOG.md
Comment (tgoyne) [inline CHANGELOG.md:14]: Incorrect merge.

File: RealmSwift/SwiftUI.swift
Comment (tgoyne) [inline RealmSwift/SwiftUI.swift:265]: Rather than duplicating all this logic in a whole bunch of places there should just be a single function that updates things which is called in each didSet.

File: RealmSwift/SwiftUI.swift
Comment (tgoyne) [inline RealmSwift/SwiftUI.swift:465]: try? is bad and this should only commit if it started the write

File: RealmSwift/Tests/ListTests.swift
Comment (tgoyne) [inline RealmSwift/Tests/ListTests.swift:304]: This should also test zero offsets and more than one.

File: CHANGELOG.md
Comment (jsflax) [reply CHANGELOG.md]: fixed

File: examples/installation/osx/swift/DynamicExample/DynamicExample.xcodeproj/project.pbxproj
Comment (tgoyne) [inline examples/installation/osx/swift/DynamicExample/DynamicExample.xcodeproj/project.pbxproj:308]: This path doesn't exist.

File: scripts/pr-ci-matrix.rb
Comment (tgoyne) [inline scripts/pr-ci-matrix.rb:50]: You also need to run this script to populate the file that jenkins actually looks at.

File: RealmSwift/EmbeddedObject.swift
Comment (jsflax) [reply RealmSwift/EmbeddedObject.swift:168]: There doesn't seem to be a clean way to do this from an extension.

Comment (tgoyne) [reply RealmSwift/EmbeddedObject.swift:168]: The public thing needs to be on Object/EmbeddedObject which then could call an internal function on the ObjectBase extension rather than putting the public thing on ObjectBase.

File: Realm/Tests/SwiftUITestHostUITests/SwiftUITestHostUITests.swift
Comment (jsflax) [reply Realm/Tests/SwiftUITestHostUITests/SwiftUITestHostUITests.swift:28]: I excluded the file using swift version guards and made the tests themselves guarded by the `available` directive.

**Comment by jsflax** [comment]
@tgoyne , addressed all comments– still haven't come up with a better name than `FetchRealmResults` though. Added a conditional in `Realm.add` that checks if SwiftUI KVO is active on an unmanaged objects and removes if so. Added a test for this as well. PTAL.

File: RealmSwift/SwiftUI.swift
Comment (tgoyne) [inline RealmSwift/SwiftUI.swift:141]: This shouldn't be asking for the new value as it doesn't use it.

File: RealmSwift/SwiftUI.swift
Comment (tgoyne) [inline RealmSwift/SwiftUI.swift:276]: Why?

File: scripts/.jenkins.yml
Comment (tgoyne) [inline scripts/.jenkins.yml:2]: This is the wrong path for this file.

**Review by tgoyne — COMMENTED** [review]
changelog merge is still incorrect

**Comment by jsflax** [comment]
I fixed the CHANGELOG (🤞), renamed `FetchRealmResults` to `ObservedResults`, renamed `KVO` to `SwiftUIKVO`, moved `.jenkins.yml` file to root, and added an `_objectSchema` method to `ObjectBase` as that had been missing.

**Comment by tgoyne** [comment]
There isn't really a reason to add _objectSchema; just using RLMObjectBaseObjectSchema() directly is basically the same code but with less copying involved.

**Comment by jsflax** [comment]
I'll revert that. I'm having a bit of difficulty understanding why the swiftly tests continue to fail– https://ci.realm.io/job/objc_pr/target=swiftpm,xcode_version=12.1/7669/console

Any thoughts on why the guards I've added aren't sufficient?

**Comment by tgoyne** [comment]
StateObject requires Xcode 12.2, which is Swift 5.3.1

**Comment by jsflax** [comment]
Screenshot 2021-02-12 at 18 38 56

@tgoyne our first passing UI tests 😄 . Let me know if you have any more thoughts

File: build.sh
Comment (tgoyne) [inline build.sh:671]: This should be in the scheme's pre-build action, not here.

File: Configuration/SwiftUITestHost.xctestplan
Comment (tgoyne) [inline Configuration/SwiftUITestHost.xctestplan:12]: We want the main thread checker enabled for this (and UI tests are an example of a case where it might even be useful). It's only disabled in the Xcode 11 version of the test plans, not the default one.

File: Realm/Tests/SwiftUITestHost/Objects.swift
Comment (tgoyne) [inline Realm/Tests/SwiftUITestHost/Objects.swift:47]:
```suggestion
    let reminders = RealmSwift.List<Reminder>()
```

File: Realm/Tests/SwiftUITestHost/SwiftUITestHostApp.swift
Comment (tgoyne) [inline Realm/Tests/SwiftUITestHost/SwiftUITestHostApp.swift:54]:
```suggestion
        .navigationBarItems(trailing: Button("Save") {
```

File: Realm/Tests/SwiftUITestHost/SwiftUITestHostApp.swift
Comment (tgoyne) [inline Realm/Tests/SwiftUITestHost/SwiftUITestHostApp.swift:70]:
```suggestion
        return newReminderAdded && list.reminders.last == reminder
```

File: RealmSwift/Object.swift
Comment (tgoyne) [inline RealmSwift/Object.swift:218]: The previous version of this wasn't great, but removing the mention of the queue isn't an improvement to it.

File: RealmSwift/SwiftUI.swift
Comment (tgoyne) [inline RealmSwift/SwiftUI.swift:43]: This appears to be something that's supposed to be unreachable. If that's correct, it should be a `precondition()` to make that clear, while if it's something the user actually can hit then it should be a bit more descriptive.

File: RealmSwift/SwiftUI.swift
Comment (tgoyne) [inline RealmSwift/SwiftUI.swift:47]: Shouldn't lastValue be updated each time the value changes? If that doesn't make sense to do then it should be `initialValue`.

File: RealmSwift/SwiftUI.swift
Comment (tgoyne) [inline RealmSwift/SwiftUI.swift:336]: Making `Realm.Configuration` conform to Equatable and putting this logic there seems better.

File: RealmSwift/SwiftUI.swift
Comment (tgoyne) [inline RealmSwift/SwiftUI.swift:444]: This logic shows up in a lot of places and would be worth extracting to a function.

File: RealmSwift/Tests/SwiftUITests.swift
Comment (tgoyne) [inline RealmSwift/Tests/SwiftUITests.swift:133]: Indentation is off in a bunch of spots after the renames.

**Review by tgoyne — COMMENTED** [review]
Took another pass over the whole thing. Mostly looks good and all my remaining comments are minor things.

File: Realm/Tests/SwiftUITestHost/Objects.swift
Comment (jsflax) [reply Realm/Tests/SwiftUITestHost/Objects.swift:47]: I've had to unapply this suggestion– to be able to use the `$` convenience methods, it needs to be mutable.

**Review by tgoyne — APPROVED** [review]

**Pull request merged**
